models/db.go

```diff
@@ -16,7 +16,7 @@ var conn *gorm.DB
 type DB interface {
```

Comment:
Really important note, thanks, fixed. What do you think about RateFetcher interface?

Reply:
you almost got it, @Hukyl. It looks great [here](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pull/2/files#diff-9d8f319f7ce4c3c7f8faba5806e81f577540e3612406164bb27d2bb53431bee9R10): you accept `RateFetcher` interface defined where it's used.

But I still have one question, why do you return [here](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pull/2/files#diff-1ca0e5e5b8a4a9cbb18c26abb6eddc193e77b6c547e9ea77a320bc78cfa56d76R78) interface instead of struct (you can make it exported)?

Reply:
you can return struct instead of interface - and it will make no difference, because your struct will implicitly implement [required interface](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pull/2/files#diff-9d8f319f7ce4c3c7f8faba5806e81f577540e3612406164bb27d2bb53431bee9R10).

`nbuRateFetcher` -> `NBURateFetcher`

```
func NewNBURateFetcher() NBURateFetcher {
	return nbuRateFetcher{}
}
```

Or am I missing something?

Reply:
Strictly speaking, yes, nothing would change, but in general, what would I do, if a new rate fetcher is created? Would I need to export an interface then?

Reply:
@Hukyl I'm sorry, but didn't get your point 🥲
Do you mean, what if you'll define one more fetcher in the same package, right?

Reply:
yes, that's right

Reply:
then just give it a different name, there shouldn't be any problem. So there would be two implementations:
1. NBURateFetcher
2. SomeOtherRateFetcher

till it satisfies the i[nterface you need](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/pull/2/files#diff-9d8f319f7ce4c3c7f8faba5806e81f577540e3612406164bb27d2bb53431bee9R10) it will work.

Or have I misunderstood the problem?

Reply:
@Hukyl maybe I explained this wrong. Here's the [official documentation](https://go.dev/wiki/CodeReviewComments#interfaces):
`Go interfaces generally belong in the package that uses values of the interface type, not the package that implements those values. The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring.`

Reply:
well, you have understood everything correctly, but, maybe I'm still coming from C++/Java interfaces implementation, there is no guarantee that SomeOtherRateFetcher will implement the interface correctly, except by the usage of this new fetcher in my `server.Client`? No formality, the new struct just exists? Maybe that's idiomatic Go, but seems a bit odd to me

Reply:
> there is no guarantee that SomeOtherRateFetcher will implement the interface correctly, except by the usage of this new fetcher in my server.Client?
yep, this is correct. This is implicit implementation 😁

Reply:
if it's not implemented correctly - you'll get an error at the compile time, or event before the compiling process, in your IDE. So, what's the difference then? :)

Reply:
Weeeeell :) Okay, I get your point, thx! I'll be sure to fix this